src/libsync/foldermetadata.h

```diff
@@ -0,0 +1,244 @@
+#pragma once
+/*
+ * Copyright (C) 2024 by Oleksandr Zolotov <alex@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "accountfwd.h"
+#include "encryptedfoldermetadatahandler.h"
+#include <csync.h>
+#include <rootencryptedfolderinfo.h>
+#include <QByteArray>
+#include <QHash>
+#include <QJsonObject>
+#include <QObject>
+#include <QSet>
+#include <QSslKey>
+#include <QString>
+#include <QVector>
+
+class QSslCertificate;
+class QJsonDocument;
+class TestClientSideEncryptionV2;
+class TestSecureFileDrop;
+namespace OCC
+{
+ // Handles parsing and altering the metadata, encryption and decryption. Setup of the instance is always asynchronouse and emits void setupComplete()
+class OWNCLOUDSYNC_EXPORT FolderMetadata : public QObject
+{
+    friend class ::TestClientSideEncryptionV2;
+    friend class ::TestSecureFileDrop;
+    Q_OBJECT
+
+    // represents a user that has access to a folder for which metadata instance is created
+    struct FolderUser {
+        QString userId;
+        QByteArray certificatePem;
+        QByteArray encryptedMetadataKey;
+    };
+
+    // based on api-version and "version" key in metadata JSON
+    enum MetadataVersion {
+        VersionUndefined = -1,
+        Version1,
+        Version1_2,
+        Version2_0,
+    };
+
+    // represents a user that has access to a specific filedrop entry of the current folder
+    struct FileDropEntryUser {
+        QString userId;
+        QByteArray decryptedFiledropKey;
+
+        inline bool isValid() const
+        {
+            return !userId.isEmpty() && !decryptedFiledropKey.isEmpty();
+        }
+    };
+
+    struct FileDropEntry {
+        QString encryptedFilename;
+        QByteArray cipherText;
+        QByteArray nonce;
+        QByteArray authenticationTag;
+        FileDropEntryUser currentUser;
+        
+        inline bool isValid() const
+        {
+            return !cipherText.isEmpty() && !nonce.isEmpty() && !authenticationTag.isEmpty() && currentUser.isValid();
+        }
+    };
+
+public:
+    struct EncryptedFile {
+        QByteArray encryptionKey;
+        QByteArray mimetype;
+        QByteArray initializationVector;
+        QByteArray authenticationTag;
+        QString encryptedFilename;
+        QString originalFilename;
+        bool isDirectory() const;
+    };
```

Comment:
why are we duplicating `SyncFileItem` but quite not ?

Reply:
@mgallien No, this is not related to SyncFileItem at all, this is an internal structure for encrypted items that are not related to sync and discovery, this is for parsing metadata, and I did not create it, it was there forever, let's just keep it please, for now, we can refactor later if absolutely need